### DISCUSSION TITLE
Protocol Design

### DISCUSSION THREAD

**Discussion by kemingy** [mosecorg/mosec#3] [RFC]
This protocol will be used for data transformation between the serving controller and workers through the Unix domain socket ([why](https://lists.freebsd.org/pipermail/freebsd-performance/2005-February/001143.html)).

* socket type: `SOCK_STREAM`
* blocking: true
* byte order: big-endian

### Protocol

**Header**

    | flag | num |

* `flag` (16-bit): different status code
  * `1`: 200 OK
  * `2`: 400
  * `4`: 422 Validation Error
  * `8`: 500 Internal Server Error
  * the rest are reserved
* `num` (16-bit): number of records in the body

**Body**

    | id_1 | length_1 | data_1 |
    | id_2 | length_2 | data_2 |
    ...
    | id_n | length_n | data_n |

* `n`: [1, `max_batch_size`], number of tasks being batched
* `id` (32-bit): task ID
* `length` (32-bit): length of this data
* `data` (0 - 2^32 bytes)

**Comment by lkevinzc** [comment]
reference:
-  https://realpython.com/python-sockets/
